### PR TITLE
parallelize summarizations, add async llm utils with retry

### PR DESCRIPTION
**Pull request by itaim**
### Background

I am using Auto-GPT search, browse and summarize code in another project and using this parallelization of single url summarization and summarizing multiple urls. 
As most of the processing is IO bound I find this approach to be useful.

### Changes

* Added ChatRequest class which encapsulates openai.ChatCompletion parameters and adds an optional key to requests.
* Added ChatResponse class which encapsulates a ChatRequest and its reply (or Exception)
* Added async_chat_completion to llm_utils with ChatRequest input ChatResponse output
* Changed summarize_text to be async, parallelizing chunks summarizations.
* Created a search file and moved relevant functions into it from commands
* Made get_text_summary async. Can be used to summarize multiple urls in parallel depending on num_sources requested (default is 1)
* Moved main loop into an async function to be asyncio.run from  main.
* Had an issue with Pinecone memory clear() so put it in try/except

### Test Plan

test_search - test_search_and_summarize_multiple_sources get 5 links from google_official_search and creates summaries from 3. It fails on the NYT and generates the third summary from the 4th url. (can't really build on this flow)

### Change Safety

- [x] I have added tests to cover my changes
- [x] I have considered potential risks and mitigations for my changes

I ran main with --continuous for a while it was stable.

### CODE REVIEW

**Review by nponeccop — CHANGES_REQUESTED** [review]
Revert the whitespace fixes. They will be merged in #590. And resolve the conflicts.

File: scripts/config.py [hunk 1]
```diff
@@ -52,12 +54,15 @@ def __init__(self):
...
             openai.api_base = self.openai_api_base
             openai.api_version = self.openai_api_version
+        else:
+            self.openai_deployment_id = None
 
         self.elevenlabs_api_key = os.getenv("ELEVENLABS_API_KEY")
 
         self.use_mac_os_tts = False
         self.use_mac_os_tts = os.getenv("USE_MAC_OS_TTS")
-
```
Comment (nponeccop) [inline scripts/config.py]: This whitespace change is unrelated.

File: scripts/main.py [hunk 2]
```diff
@@ -293,118 +295,122 @@ def parse_arguments():
...
+            full_message_history.append(chat.create_chat_message("system", result))
+            print_to_console("SYSTEM: ", Fore.YELLOW, result)
+        else:
+            full_message_history.append(
+                chat.create_chat_message(
+                    "system", "Unable to execute command"))
+            print_to_console("SYSTEM: ", Fore.YELLOW, "Unable to execute command")
+
+if __name__ == "__main__":
+    asyncio.run(main_loop())
```
Comment (nponeccop) [inline scripts/main.py]: The final CRLF is missing

File: scripts/config.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import abc
 import os
+
```
Comment (nponeccop) [inline scripts/config.py]: Revert this

File: scripts/config.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
...
 import openai
 from dotenv import load_dotenv
+
```
Comment (nponeccop) [inline scripts/config.py]: And this. While it's a good fix it's unrelated to your PR

**Review by nponeccop — CHANGES_REQUESTED** [review]

File: scripts/config.py (hunk 1 above)
Comment (itaim) [reply scripts/config.py]: @nponeccop, consider using black for formatting/linting. could make life easier with so many PRs to review.

**Comment by nponeccop** [comment]
@itaim There are conflicts now

**Comment by Pwuts** [comment]
If this PR is still relevant, please resolve the conflicts so we can review

**Comment by Pwuts** [comment]
Related: #3222

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by Pwuts** [comment]
@collijk or @ntindle can you check if any of this can be (partially) cherry-picked into re-arch?

**Comment by Pwuts** [comment]
Most of this has been implemented in what is now `forge.llm.providers`, which is used throughout `autogpt`.

Parallel summarization of chunks is currently *not* implemented because on large bodies of text it would hit rate limits of the completion API. `forge.llm.providers` would need a mechanism to configure & limit the number of parallel requests.

**Comment by Pwuts** [comment]
Thank you for submitting this pull request and for your interest in the project!
